dbms/src/Storages/DeltaMerge/SSTFilesToBlockInputStream.cpp

```diff
@@ -64,20 +64,97 @@ SSTFilesToBlockInputStream::~SSTFilesToBlockInputStream() = default;
 
 void SSTFilesToBlockInputStream::readPrefix()
 {
-    for (UInt64 i = 0; i < snaps.len; ++i)
+    bool use_multi = false;
+    if (snaps.len > 3)
     {
-        const auto & snapshot = snaps.views[i];
-        switch (snapshot.type)
+        use_multi = true;
+    }
+    else
+    {
+        int c[3] = {0};
+        for (UInt64 i = 0; i < snaps.len; ++i)
         {
-        case ColumnFamilyType::Default:
-            default_cf_reader = std::make_unique<SSTReader>(proxy_helper, snapshot);
-            break;
-        case ColumnFamilyType::Write:
-            write_cf_reader = std::make_unique<SSTReader>(proxy_helper, snapshot);
-            break;
-        case ColumnFamilyType::Lock:
-            lock_cf_reader = std::make_unique<SSTReader>(proxy_helper, snapshot);
-            break;
+            const auto & snapshot = snaps.views[i];
+            auto t = static_cast<std::underlying_type_t<ColumnFamilyType>>(snapshot.type);
+            c[t]++;
+            if (c[t] > 1)
+            {
+                use_multi = true;
+                break;
+            }
+        }
+    }
+    if (use_multi)
+    {
+        LOG_FMT_INFO(log, "Read SST Files with MultiSSTReader {}", this->region->id());
+        std::vector<SSTView> ssts;
+        bool flag = false;
+        ColumnFamilyType prev_type;
+        int size_write = 0;
+        int size_lock = 0;
+        int size_default = 0;
+
+        auto make_inner_func = [&](const TiFlashRaftProxyHelper * proxy_helper, SSTView snap) {
+            return std::make_unique<SSTReader>(proxy_helper, snap);
+        };
+        auto generate_cf_reader = [&]() {
+            // Generate old cf reader.
+            switch (prev_type)
+            {
+            case ColumnFamilyType::Default:
+                size_default = ssts.size();
```

Comment:
The order is ensured by using exactly the interface `CfFile::file_paths()` in proxy.

Reply:
I vote for map+vector or 1 vector for each cf, then we don't rely on the assumption of TiKV. Cause it's hard to constrain tikv behavior change by tiflash.
And suggest adding an exception when tiflash meets unknown cf.

Reply:
I am confused what the problem is? @breezewish @JaySon-Huang Is the case
```
default 1
lock 1
default 2
lock 2
```
rather than the case
```
default 2
default 1
```
you all concern?
Then proxy will guarantee an order. However three vectors is also acceptable.

Reply:
The point is can we be aware of TiKV behavior changes in the early stage (like internal testing). Or automatically adapt to the behavior changes.

Reply:
@CalvinNeo  Yeah, I'm indeed concerning about some day this case may happen:

```
default 1
lock 1
default 2
lock 2
```

It would be better if we could be (even) compatible with this kind of order, or we could have some checks to detect it.

In the current impl, when this kind of order happens, seems that we will silently lose some data. Then it may be a hard work for us to troubleshoot.

Reply:
For the first question, though proxy prevent this situation, I have adopt the map+vector(actually 3 vectors) idea in my later commits in TiFlash, so there is a double ensurance.
For the second question, we have no means to validate if a list of file is in order. The order is totally a blackbox for us. We use what TiKV gives us.